File: factors.py
import argparse


def factors(num):
    i = 2
    current = num

    factors = {}

    while i <= num / 2:
        if current % i == 0:
            if i in factors:
                factors[i] += 1
            else:
                factors[i] = 1
            current = current / i
        else:
            i += 1

    return factors


def printOutput(factors):
    out = []
    for key, value in factors.items():
        if value == 1:
            out.append(f"{key}")
        else:
            out.append(f"{key}^{value}")

    return ", ".join(out)


if __name__ == "__main__":

    parser = argparse.ArgumentParser()
    parser.add_argument("num", type=int, help="Number to calculate factors")
    args = parser.parse_args()

    divs = factors(args.num)
    if len(divs) == 0:
        print(f"{args.num} is a prime number")
    else:
        print(f"factors for {args.num} are: ({printOutput(divs)})")

File: factors.js
'use strict';

if (process.argv.length !== 3) {
    console.log('Usage: node factors.js <number>');
    process.exit(1);
}

function calcFactors(number) {
    const factorsHash = {};
    let current = number
    let i = 2
    while (i <= number / 2) {
        if (current % i === 0) {
            current = current / i;
            if (factorsHash[i]) {
                factorsHash[i] += 1;
            } else {
                factorsHash[i] = 1;
            }
        } else {
            i++;
        }
    }
    return factorsHash;
}

function printFactors(factors) {
    const result = [];
    for (const key in factors) {
        if (factors[key] > 1) {
            result.push(`${key}^${factors[key]}`)
        } else {
            result.push(`${key}`)
        }
    }
    return result.join(', ')
}

const num = process.argv[2]
const res = calcFactors(num)
if (Object.keys(res).length === 0) {
    console.log(`${num} is a prime number`)
} else {
    console.log(`factors for ${num} are: (${printFactors(res)})`)
}

